File: src/02_protobuffer/main/main.cc
#include <iostream>
#include <fstream>
#include <string>
#include <addressbook/addressbook.pb.h>
using namespace std;

static void CreateAddressBook(addressbook::AddressBook *address_book) {
  for (int i=0; i<3; i++) {
    addressbook::Person *person = address_book->add_person();
    person->set_id(i);
    *person->mutable_name() = string("John-")+ std::to_string(i);
    person->set_email(string("JohnMail-")+ std::to_string(i));
    for (int j=0; j<3; j++) {
      addressbook::Person::PhoneNumber* phone_number = person->add_phones();
      phone_number->set_number(string("JohnNumber-")+ std::to_string(i) + std::to_string(j));
      phone_number->set_type(static_cast<addressbook::Person_PhoneType>(j%3));
    }
  }
}

// Main function:  Reads the entire address book from a file,
//   adds one person based on user input, then writes it back out to the same
//   file.
int main(int argc, char* argv[]) {
  // Verify that the version of the library that we linked against is
  // compatible with the version of the headers we compiled against.
  GOOGLE_PROTOBUF_VERIFY_VERSION;

  if (argc != 2) {
    cerr << "Usage:  " << argv[0] << " ADDRESS_BOOK_FILE" << endl;
    return -1;
  }

  //read & unmarshal
  addressbook::AddressBook address_book;
  {
    // Read the existing address book.
    fstream input(argv[1], ios::in | ios::binary);
    if (!input) {
      cout << argv[1] << ": File not found.  Creating a new file." << endl;
      CreateAddressBook(&address_book);
    } else if (!address_book.ParseFromIstream(&input)) {
      cerr << "Failed to parse address book." << endl;
      return -1;
    }
  }
  
  //dump
  for (int i=0; i< address_book.person_size(); i++) {
    const addressbook::Person &person = address_book.person(i);
    printf("\n%d\n", i);
    printf("\tid:%d\n", person.id());
    printf("\tname:%s\n", person.name().c_str());
    printf("\temail:%s\n", person.email().c_str());
    for (int j=0; j<person.phones_size(); j++) {
      const addressbook::Person::PhoneNumber& phone_number = person.phones(j);
      printf("\t\tnumber:%s\n", phone_number.number().c_str());
      printf("\t\ttype:%d\n", phone_number.type());
    }
  }

  //marshal & write 
  {
    // Write the new address book back to disk.
    fstream output(argv[1], ios::out | ios::trunc | ios::binary);
    if (!address_book.SerializeToOstream(&output)) {
      cerr << "Failed to write address book." << endl;
      return -1;
    }
  }

  // Optional:  Delete all global objects allocated by libprotobuf.
  google::protobuf::ShutdownProtobufLibrary();
  return 0;
}

File: src/02_protobuffer/main/main.go
package main

import(
  _"fmt"
  pb "addressbook"
  "strconv"
  "io/ioutil"
  "log"
  "os"
  "github.com/golang/protobuf/proto"
)


func CreateAddressBook(ab *pb.AddressBook, n int)  {
  var i int32;
  for i=0; i<3; i++ {
    person := &pb.Person{
        Id:    i,
        Name:  "John-"+strconv.Itoa(int(i)),
        Email: "jdoe@example.com",
    };

    for j:=0; j<3; j++ {
      person.Phones = append(person.Phones,
        &pb.Person_PhoneNumber{Number: "12345678#"+strconv.Itoa(int(j)), Type: pb.Person_PhoneType(j%3)},
      );
    }

    ab.Person = append(ab.Person, person);
  }
  log.Printf("%v\n", len(ab.Person));
}

type AddressBookLoader struct {
  
};

func NewAddressBookLoader() *AddressBookLoader {
  return &AddressBookLoader {};
}

func (p *AddressBookLoader) Unmarshal(in []byte, ab *pb.AddressBook) error {
  var err error;
  if err = proto.Unmarshal(in, ab); err != nil {
    log.Printf("failed to unmarshal address book: %v\n", err);
  }
  return err;
}

func (p *AddressBookLoader) ReadFromFile(filename string, ab *pb.AddressBook) error {
  var in []byte;
  var err error;

  // Read the existing address book.
  in, err = ioutil.ReadFile(filename);
  if err != nil {
    if os.IsNotExist(err) {
      log.Printf("failed to read file (FileNotFound) %v \n", filename);
    } else {
      log.Printf("failed to read file: %v", err);
    }
    return err;
  }

  return p.Unmarshal(in, ab);
}

func (p *AddressBookLoader) Marshal(ab *pb.AddressBook) ([]byte, error) {
  return proto.Marshal(ab);
}

func (p *AddressBookLoader) WriteToFile(filename string, ab *pb.AddressBook) error {
  var buf []byte;
  var err error = nil;
  
  buf, err = p.Marshal(ab);
  if err != nil {
    return err;
  }

  if err = ioutil.WriteFile(filename, buf, 0644); err != nil {
    log.Fatalln("Failed to write address book:", err);
  }
  return err;
}

func main() {
  ab := &pb.AddressBook{};
  loader := NewAddressBookLoader();
  var filename string = "";

  if len(os.Args) != 2 {
    log.Printf("usage: %v filename\n", os.Args[0]);
    return;
  }

  filename = os.Args[1];

  //read
  if err := loader.ReadFromFile(filename, ab); err != nil {
    log.Printf("create new one\n");
    CreateAddressBook(ab, 10);
  } else {
    log.Printf("file read %v ok\n", filename);
  }

  //dump
  for i:=0; i<len(ab.Person); i++ {
    log.Printf("\n\n%d\n", i);
    log.Printf("\t%v %v\n", ab.Person[i].GetId(), ab.GetPerson()[i]);
  }

  //write
  if err := loader.WriteToFile(filename, ab); err != nil {
    log.Printf("failed to loader.WriteToFile %v\n", err);
  }
}
